### PR TITLE
Fixes #1031 - Show structure collapse warning when deploying and moving.

### LINKED ISSUE #1031: Confirm Deployment onto Building that will Collapse
### Environment
Megamek 0.44.0
MacOS 10.13.5
Java 8 Update 171?

### Description
When deploying a 'mech onto a building that will collapse, it would be great to have a notification that it might happen.

### LINKED ISSUE #2595: Warn before collapsing buildings via movement due to weight
As suggested on Slack, a warning popup when attempting to move through a building hex that will collapse due to the weight of the mech would be good. 

 Additionally, a warning when ending your turn on a building hex with a CF X above your mech, where X is configurable in preferences, would also be good.

### PR DESCRIPTION
**Pull request by HoneySkull**
Fixes #1031 
Fixes #2595 

Implements a feature to display a warning symbol on hexes with structures during the Deployment and Movement where the CF for the building is less than the weight of the selected entity.  

In the deployment phase, hexes that are valid deployment hexes with buildings will display a warning symbol if CF of the structure is less than the weight of the selected entity combined with any existing entities on that hex.

In the movement phase, a warning symbol will be displayed on hexes with structures that could collapse within a generous movement radius.  (max run/jump hexes in radial distance).

image

image

The structural warning symbols can be turned on an off in the view menu:
image

### CODE REVIEW

**Comment by HoneySkull** [comment]
Set as draft - I have an edge case to solve for flying units over buildings.

File: megamek/src/megamek/client/ui/swing/boardview/ConstructionWarningSprite.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+ *  and movement phase that will collapse if the entity lands-on 
+ *  or is deployed on that structure.
+ * 
+ *  From TW: If a units tonnage exceeds the CF of a building
+ *  or bridge, it will collapse.  (Or the sum of tonnage of stacked
+ *  units if multiple units occupy the hex)
+ */
+public class ConstructionWarningSprite extends HexSprite {
+
+	private static final GUIPreferences GUIP = GUIPreferences.getInstance();
```
Comment (neoancient) [inline megamek/src/megamek/client/ui/swing/boardview/ConstructionWarningSprite.java:41]: Extra indent

File: megamek/src/megamek/common/ConstructionFactorWarning.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        List<Coords> warnList = new ArrayList<Coords>();
+
+        try {
+            //Only calculate CF Warnings for entity types in the whitelist.
+            if (!entityTypeIsInWhiteList(e)) {
+                return warnList;
+            }
+
+            Coords pos = e.getPosition();
+            int range = (e.getJumpMP() > e.getRunMP()) ? e.getJumpMP() : e.getRunMP();
```
Comment (neoancient) [inline megamek/src/megamek/common/ConstructionFactorWarning.java:96]: I find this more readable, but I consider that preference.
```suggestion
            int range = Math.max(e.getJumpMP(), e.getRunMP());
```

File: megamek/src/megamek/common/ConstructionFactorWarning.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+ *
+ * MegaMek is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with MegaMek. If not, see <http://www.gnu.org/licenses/>.
+ */
+package megamek.common;
```
Comment (neoancient) [inline megamek/src/megamek/common/ConstructionFactorWarning.java:19]: Is there a reason for putting this in `megamek.common`? It depends on `megamek.client` and is not used by anything outside that package as far as I can see. Ideally client and server depend on common, but not the other way around. The current situation is pretty bad, but best not to make it worse.

**Review by neoancient — CHANGES_REQUESTED** [review]
Good overall. Just a couple quick and easy changes.

File: megamek/src/megamek/common/ConstructionFactorWarning.java (hunk 3 above)
Comment (HoneySkull) [reply megamek/src/megamek/common/ConstructionFactorWarning.java:19]: Agreed - this kind of evolved as an attempt to consolidate the core logic for this feature into a single unit testable place.  Is it alright if I move it to `package megamek.client.ui.swing;`

File: megamek/src/megamek/common/ConstructionFactorWarning.java (hunk 2 above)
Comment (HoneySkull) [reply megamek/src/megamek/common/ConstructionFactorWarning.java:96]: Of course! If I wasn't drinking "eggnog", I would have thought of this.

File: megamek/src/megamek/client/ui/swing/boardview/ConstructionWarningSprite.java (hunk 1 above)
Comment (HoneySkull) [reply megamek/src/megamek/client/ui/swing/boardview/ConstructionWarningSprite.java:41]: Missed it - will fix.

Comment (HoneySkull) [reply megamek/src/megamek/client/ui/swing/boardview/ConstructionWarningSprite.java]: Updated.

File: megamek/src/megamek/common/ConstructionFactorWarning.java (hunk 2 above)
Comment (HoneySkull) [reply megamek/src/megamek/common/ConstructionFactorWarning.java]: Updated.

File: megamek/src/megamek/common/ConstructionFactorWarning.java (hunk 3 above)
Comment (HoneySkull) [reply megamek/src/megamek/common/ConstructionFactorWarning.java]: Moved to `megamek.client.ui.swing`.

**Review by neoancient — APPROVED** [review]

**Pull request merged**